File: hackerrank/30 days of code/day 12, inheritance/inheritance.cpp
// https://www.hackerrank.com/challenges/30-inheritance/problem?isFullScreen=true
// From C++23 onward: ranges::fold_left_first

#include <algorithm>
#include <iostream>
#include <iterator>
#include <vector>

using namespace std;

struct personal_data {
    string first_name, last_name;
    int id{};
};

struct student_data {
    vector<int> scores;
};

class Person {
    string first_name, last_name;
    int id;

public:
    explicit Person(const personal_data& data);
    void print_person() const;
};

class Student : public Person {
    vector<int> scores;

public:
    explicit Student(const personal_data& personal_d, const student_data& student_d);
    [[nodiscard]] char calculate_grade() const;
};

Person::Person(const personal_data& data)
    : first_name{data.first_name}, last_name{data.last_name}, id{data.id}
{
}

void Person::print_person() const
{
    cout << "Name: " << last_name << ", " << first_name << '\n';
    cout << "ID: " << id << '\n';
}

Student::Student(const personal_data& personal_d, const student_data& student_d)
    : Person{personal_d}, scores{student_d.scores}
{
}

char Student::calculate_grade() const
{
    if (auto avg{*ranges::fold_left_first(scores, plus{}) / scores.size()}; avg > 100 || avg < 40)
        return 'T';
    else if (avg >= 90)
        return 'O';
    else if (avg >= 80)
        return 'E';
    else if (avg >= 70)
        return 'A';
    else if (avg >= 55)
        return 'P';
    return 'D';
}

//////////////////////////////////////////////////

tuple<personal_data, student_data> read_input();

int main()
{
    auto [personal_d, student_d]{read_input()};
    Student student{personal_d, student_d};
    student.print_person();
    cout << "Grade: " << student.calculate_grade();
    return 0;
}

tuple<personal_data, student_data> read_input()
{
    personal_data personal_d;
    student_data student_d;
    int n_scores;
    cin >> personal_d.first_name >> personal_d.last_name >> personal_d.id >> n_scores;
    student_d.scores.resize(n_scores);
    copy_n(istream_iterator<int>(cin), n_scores, student_d.scores.begin());
    return {personal_d, student_d};
}

File: hackerrank/30 days of code/day 12, inheritance/inheritance.java
// https://www.hackerrank.com/challenges/30-inheritance/problem?isFullScreen=true

import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public class Solution {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        final String firstName = scan.next();
        final String lastName = scan.next();
        final int id = scan.nextInt();
        final int nScores = scan.nextInt();

        final List<Integer> scores = Arrays.asList(new Integer[nScores])
                                     .stream()
                                     .map(x -> scan.nextInt())
                                     .collect(Collectors.toList());

        final Student student = new Student(firstName, lastName, id, scores);
        student.printPerson();
        System.out.println("Grade: " + student.calculate());
    }
}

    class Person {
        protected String firstName, lastName;
        protected int id;

        Person(String firstName, String lastName, int id) {
            this.firstName = firstName;
            this.lastName = lastName;
            this.id = id;
        }

        public void printPerson() {
            System.out.println("Name: " + lastName + ", " + firstName +   "\nID: " + id);
        }
    }

        class Student extends Person {
            private List<Integer> scores;

            public Student(String firstName, String lastName, int id, List<Integer> scores) {
                super(firstName, lastName, id);
                this.scores = scores;
            }

            public char calculate() {
                final int avg = (int)scores.stream().mapToDouble(i -> i).average().orElse(0.0);

                if (avg >= 90 && avg <= 100)
                    return 'O';
                else if (avg >= 80 && avg < 90)
                    return 'E';
                else if (avg >= 70 && avg < 80)
                    return 'A';
                else if (avg >= 55 && avg < 70)
                    return 'P';
                else if (avg >= 40 && avg < 55)
                    return 'D';
                else
                    return 'T';
            }
        }
